accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/QueryHelper.java

```diff
@@ -73,32 +77,47 @@
                     }
                 }
             } catch (RepositoryException e) {
-                AcHelper.LOG.error("Exception: {}", e);
+                LOG.error("Exception: {}", e);
             }
         }
         Set<Node> nodes = new LinkedHashSet<Node>();
         try {
             // get the rep:policy node of "/", if existing
-            if (session.nodeExists("/rep:policy")) {
-                nodes.add(session.getNode("/rep:policy"));
+            if (session.nodeExists(ROOT_REP_POLICY_NODE)) {
+                nodes.add(session.getNode(ROOT_REP_POLICY_NODE));
             }
-            if (session.nodeExists("/" + Constants.REPO_POLICY_NODE)) {
-                nodes.add(session.getNode("/" + Constants.REPO_POLICY_NODE));
+            if (session.nodeExists(ROOT_REPO_POLICY_NODE)) {
+                nodes.add(session.getNode(ROOT_REPO_POLICY_NODE));
             }
 
             // get the rep:policy node of "/home", if existing
-            if (session.nodeExists("/home/rep:policy")) {
-                nodes.add(session.getNode("/home/rep:policy"));
+            if (session.nodeExists(HOME_REP_POLICY)) {
+                nodes.add(session.getNode(HOME_REP_POLICY));
             }
 
+            boolean indexForRepACLExists = session.nodeExists(OAK_INDEX_PATH_REP_ACL);
+            LOG.debug("Index for repACL exists: {}",indexForRepACLExists);
+            String queryForAClNodes = indexForRepACLExists ? 
+                    "SELECT * FROM [rep:ACL] WHERE ISDESCENDANTNODE([%s])" : 
+                    "SELECT ace.* FROM [rep:ACE] AS ace WHERE ace.[rep:principalName] IS NOT NULL AND ISDESCENDANTNODE(ace, [%s])";
+            LOG.debug("Query to obtain all ACLs: {}", queryForAClNodes);
+            
             for (String path : paths) {
-                String query = "SELECT * FROM [rep:ACL] WHERE ISDESCENDANTNODE([" + path + "])";
-                nodes.addAll(QueryHelper.getNodes(session, query, Query.JCR_SQL2));
+                if(StringUtils.equals(path, ROOT_REP_POLICY_NODE) || StringUtils.equals(path, ROOT_REPO_POLICY_NODE)) {
+                    continue;
+                }
+                
+                String query = String.format(queryForAClNodes, path);
+
+                long startTime1 = System.currentTimeMillis();
+                Set<Node> nodesResult = indexForRepACLExists ? 
+                        getNodes(session, query, Query.JCR_SQL2):
+                        getDistinctParentNodes(session, query, Query.JCR_SQL2);
+                LOG.debug("Query to find ACLs under {} ran in {}ms (count ACLs: {})", path, System.currentTimeMillis()-startTime1, nodesResult.size());
```

Comment:
Performance tests with an instance with 2 Mio pages and a fair amount of ACLs in them looked ok (still under 1sec with this approach, >100ms with the old approach)

Reply:
the test above was with a publish instance (and 2 Mio pages), it turns out that on an author instance with the same content the performance can degrade to around 8 seconds. I think this is still acceptable (also because for the case the index is there, it works with the old speed, see https://github.com/Netcentric/accesscontroltool/pull/392#discussion_r341921557)